все переписано на mongodb

sadhaze opened the pull request.

sadhaze committed:
все переписано на mongodb

stvort commented on pom.xml:
```diff
@@ -23,27 +23,31 @@
+<!--        <dependency>
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-data-jpa</artifactId>
             <version>2.0.3.RELEASE</version>
```
Лучше не хранить не используемый код

stvort commented on src/main/java/otus/library/domain/Book.java:
```diff
@@ -1,28 +1,17 @@
     private Author author;
 
-    @ManyToOne
     private Genre genre;
```
Автор и жанр это тоже документы для каждого из которых есть своя коллекция. Как планируется обеспечить ссылочную целостность при удалении, например, автора? Да. Сейчас для удаления нет команды шелл, но при этом в репозитории такой метод есть). Чтобы не дублировать хранение авторов и жанров можно воспользоваться ```@DBRef```

stvort commented on src/main/java/otus/library/domain/Author.java:
```diff
@@ -1,12 +1,9 @@
+@Document
 public class Author {
     @Id
     private Long id;
```
Я так понял, что сейчас пользователь должен сам вводить id. Лучше так не делать) Можно или генерировать id самостоятельно, например, через создание коллекции, хранящей последнее значение счетчика, либо предоставить это дело монге, сделав id строкой (на самом деле ObjectId).

stvort commented on src/main/java/otus/library/service/ShellUi.java:
```diff
@@ -30,13 +30,13 @@ private ShellUi(
     @ShellMethod(value = "Show author by ID")
     public String getAuthor(Long id){
-        Author author = authorRepository.getOne(id);
+        Author author = authorRepository.findById(id).get();
```
Не самое правильное использование Optional (тут и далее). Лучше воспользоваться такими преимуществами, как orElseGet или orElseThrow, ifPresent

stvort commented on src/test/java/otus/library/repository/AuthorRepositoryTest.java:
```diff
@@ -1,20 +1,36 @@
-@DataJpaTest
+import java.util.Iterator;
+
+@SpringBootTest(properties = {
```
Репозитории монги лучше тестировать через ```@DataMongoTest```

stvort commented on src/test/java/otus/library/repository/AuthorRepositoryTest.java:
```diff
@@ -1,20 +1,36 @@
     AuthorRepository authorRepository;
 
+    @BeforeEach
+    void dbWiper(){
```
Луче просто тестировать на встроенной монге (flapdoodle) + ```@DirtiesContext```

sadhaze committed:
1) Удалил неиспользуемые зависимости
2) Добавил генерацию id
3) Добавил orElseTrow для Optional
4) Переделал тесты на embedded mongo

sadhaze committed:
1) Исправил тесты
2) Добавил 2 листнера на удаление
